In [1]:
import pandas as pd
pd.set_option("display.max_columns", 101)

# Clean and prepare dataset

We create a dataframe named df_pivot that contains job features (skills and knowledge) as columns and job titles as rows. For the purpose of this project, jobs are filtered depending on their job zones. 
All the data can be found here:
https://www.onetcenter.org/database.html

In [2]:
# Load datasets into Pandas DataFrame
skills = pd.read_excel('../databases/ONET/Skills.xlsx')

In [3]:
knowledge = pd.read_excel('../databases/ONET/Knowledge.xlsx')

In [4]:
# Merge skills and knowledge files
df = skills.append(knowledge)

In [5]:
# Filter per job zones 3, 4 and 5
job_zones = pd.read_excel('../databases/ONET/Job Zones.xlsx')
jobzone345 = job_zones[job_zones['Job Zone'].isin([3,4,5])]['O*NET-SOC Code'].tolist()
df = df[df['O*NET-SOC Code'].isin(jobzone345)]

In [6]:
# Take out unnecessary columns
cols_to_keep = [
    'O*NET-SOC Code', 
    'Title', 
    'Element Name', 
    'Scale ID',
    'Data Value'
]

df = df[cols_to_keep].copy()

In [7]:
# Merge columns 'Element Name' and 'Scale ID' into 'element_scale_id'
df['element_scale_id'] = df['Element Name'] + '_' + df['Scale ID'] 

In [8]:
# Pivot table
df_pivot = df.pivot_table(index=['Title'], columns='element_scale_id', values='Data Value')

In [9]:
df_pivot.head()

element_scale_id                        Active Learning_IM  \
Title                                                        
Accountants                                           3.25   
Actuaries                                             3.38   
Acupuncturists                                        3.12   
Acute Care Nurses                                     3.75   
Adapted Physical Education Specialists                3.50   

element_scale_id                        Active Learning_LV  \
Title                                                        
Accountants                                           3.38   
Actuaries                                             4.12   
Acupuncturists                                        3.50   
Acute Care Nurses                                     3.88   
Adapted Physical Education Specialists                3.62   

element_scale_id                        Active Listening_IM  \
Title                                                         
Accountants                                            3.88   
Actuaries                                              4.00   
Acupuncturists                                         3.75   
Acute Care Nurses                                      4.00   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Active Listening_LV  \
Title                                                         
Accountants                                            3.62   
Actuaries                                              4.12   
Acupuncturists                                         3.62   
Acute Care Nurses                                      3.88   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Administration and Management_IM  \
Title                                                                      
Accountants                                                         3.43   
Actuaries                                                           3.31   
Acupuncturists                                                      2.67   
Acute Care Nurses                                                   2.89   
Adapted Physical Education Specialists                              2.44   

element_scale_id                        Administration and Management_LV  \
Title                                                                      
Accountants                                                         4.00   
Actuaries                                                           4.72   
Acupuncturists                                                      2.57   
Acute Care Nurses                                                   2.96   
Adapted Physical Education Specialists                              2.56   

element_scale_id                        Biology_IM  Biology_LV  \
Title                                                            
Accountants                                   1.02        0.03   
Actuaries                                     1.16        0.47   
Acupuncturists                                3.45        4.05   
Acute Care Nurses                             3.32        3.50   
Adapted Physical Education Specialists        2.65        2.81   

element_scale_id                        Building and Construction_IM  \
Title                                                                  
Accountants                                                     1.21   
Actuaries                                                       1.25   
Acupuncturists                                                  1.60   
Acute Care Nurses                                               1.18   
Adapted Physical Education Specialists                          1.58   

element_scale_id                        Building and Construction_LV  \
Title                                                                  
Accountants                                                     0.40

In [10]:
# Display size of the matrix
df_pivot.shape

(640, 134)

Let's factorize this matrix to (640, 2)

# Create the embedding with UMAP

We create an embedding with UMAP and project it in 2D.

In [11]:
import umap

In [12]:
# Create embedding
embedding = umap.UMAP(n_neighbors=15, n_components=2, random_state=42,
                      min_dist=0.3,
                      metric='correlation').fit_transform(df_pivot.values)

In [13]:
# Display shape of embedding
embedding.shape

(640, 2)

In [14]:
embedding

array([[ 1.5840942 , -4.695757  ],
       [ 1.1996833 , -4.093121  ],
       [ 3.6107712 ,  0.3655267 ],
       ...,
       [-1.7982806 , -2.4191442 ],
       [-6.2889442 , -4.4923296 ],
       [-0.7414811 , -0.30035707]], dtype=float32)

We now have an array of floats that correspond to the coordinates of each point of the dataset in the target space. Let's visualize it.

# Visualization with plotly

In [15]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)
import plotly.graph_objs as go

In [16]:
data = [
    go.Scatter(
        x=embedding[:,0],
        y=embedding[:,1],
        mode='markers',
        text=df_pivot.index.tolist(),
        opacity=0.5
    )
]
layout = go.Layout(
    title='Jobs similarity map (based on skills and knowledge) with UMAP',
    hovermode='closest'
)
fig = go.Figure(data=data, layout=layout)
plot_url = iplot(fig, filename='hover-chart-basic')

Thanks!